File: Lab_2/DDL.c
#include <stdio.h>
#include <stdlib.h>

struct SLL{
    int data;
    struct SLL *next, *prev;
};

struct SLL *first, *last = NULL;

void display(){
    if(first == NULL)
        printf("Empty List.\n");
    else{
        struct SLL *temp = first;
        while(temp!=NULL){
            printf("%d -> ",temp->data);
            temp = temp->next;   
        }
        printf("NULL\n");
    }
}

void insert_at_beginning(int element){
    struct SLL *NewNode;
    NewNode = (struct SLL*)malloc(sizeof(struct SLL));
    
    NewNode->data = element;
    NewNode->next = NULL;
    NewNode->prev = NULL;

    if(first == NULL)
        first = last = NewNode;
    else
    {
        NewNode->next = first;
        first->prev = NewNode;
        first = NewNode;
    }
    printf("%d WAS INSERTED at beginning.\n", first->data);
}

void insert_at_end(int element){
    struct SLL *NewNode;
    NewNode = (struct SLL*)malloc(sizeof(struct SLL));
    
    NewNode->data = element;
    NewNode->next = NULL;

    if(first == NULL)
        first = last = NewNode;
    else
    {
        NewNode->prev = last;
        last->next = NewNode;
        last = NewNode;
    }
    printf("%d WAS INSERTED at end.\n", last->data);
}

void insert_at_specificPostion(int element, int POS){
    struct SLL *NewNode, *temp;
    int i;
    NewNode = (struct SLL*)malloc(sizeof(struct SLL));
    
    NewNode->data = element;
    NewNode->next = NULL;

    if(POS <= 0){
        printf("Invalid Position.\n");
    }else if(POS == 1 || first == NULL){
        insert_at_beginning(element);
    }else{
        struct SLL *temp = first;

        for(i = 1; (i<POS-1) && (temp->next != NULL); i++){
            temp = temp->next;
        }

        if(temp == last)
            insert_at_end(element);
        else{
            NewNode->next = temp->next;
            NewNode->prev = temp;
            (NewNode->next)->prev = NewNode;
            temp->next = NewNode;
            printf("%d WAS INSERTED at %d.\n", NewNode->data, POS);
        }
    }
}

void delete_from_beginning(){
    struct SLL *temp;

    if(first == NULL)
        printf("ERROR: List is empty.\n");
    else if(first->next == NULL){
        temp = first;
        first = last = NULL;
        free(temp);
    }else{
        temp = first;
        first = first->next;
        free(temp);
        first->prev = NULL;
    }
    printf("Data was deleted from beginning.\n");
}

void delete_from_end(){
    struct SLL *temp;

    if(first == NULL)
        printf("ERROR: List is empty.\n");
    else if(first->next == NULL){
        temp = first;
        first = last = NULL;
        free(temp);
    }else{
        temp = last->prev;
        last = temp;
        free(temp->next);
        last->next = NULL;
    }
    printf("Data was deleted from end.\n");
}

void delete_from_specific_position(int POS){
    struct SLL *temp, *tempp;
    int i;
    if( first == NULL)
        printf("List is empty\n");
    else if(POS <= 0)
        printf("Invalid position.");
    else if(POS == 1)
        delete_from_beginning();
    else{
        temp = first;
        for(i=1; (i<POS-1) && (temp->next != last); i++)
            temp = temp->next;
        
        if(temp->next == last)
            delete_from_end();
        else{
            tempp = temp->next;
            temp->next = tempp->next;
            (tempp->next)->prev = temp;
            free(tempp);
        }

        printf("Data was deleted from %d\n", POS);
    }
}

int main(){
    insert_at_end(20);
    display();
    insert_at_end(10);
    display();
    insert_at_end(25);
    display();
    insert_at_specificPostion(30, 1);
    display();
    insert_at_specificPostion(11, 4);
    display();
    insert_at_specificPostion(200, 6);
    display();
    insert_at_specificPostion(500, 20);
    display();

    delete_from_beginning();
    display();
    delete_from_end();
    display();
    delete_from_specific_position(3);
    display();
}

File: Lab_2/DDL.py

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None

class DLL:
    def __init__(self):
        self.first = None
        self.last = None

    def display(self):
        if self.first is None:
            print("Empty List.")
        else:
            temp = self.first
            data = []
            while temp is not None:
                data.append(temp.data)
                temp = temp.next    
            print(data)
    
    def insert_at_beginning(self, element):
        NewNode = Node(element)

        if self.first is None:
            self.first = NewNode
            self.last = NewNode
        else:
            NewNode.next = self.first
            self.first.prev = NewNode
            self.first = NewNode
        
        print(f"{self.first.data} was inserted at beginning.")
    
    def insert_at_end(self, element):
        NewNode = Node(element)

        if self.first is None:
            self.first = NewNode
            self.last = NewNode
        else:
            NewNode.prev = self.last
            self.last.next = NewNode
            self.last = NewNode
        
        print(f"{self.last.data} was inserted at end.")

    def insert_at_specific_position(self, element, pos):
        NewNode = Node(element)

        if pos <= 0:
            print("[Error]: Invalid Position")
        elif pos == 1:
            self.insert_at_beginning(element)
        else:
            temp = self.first

            for i in range(1, pos-1):
                if temp.next is None:
                    break
                temp = temp.next
            
            if temp == self.last:
                self.insert_at_end(element)
            else:
                NewNode.next = temp.next
                NewNode.prev = temp
                NewNode.next.prev = NewNode
                temp.next = NewNode
                print(f"{NewNode.data} was inserted at {pos}")
    
    def delete_from_beginning(self):
        if self.first is None:
            print("[Error]: List is empty.")
        elif self.first.next is None:
            self.first = self.last = None
        else:
            self.first = self.first.next
            self.first.prev = None
        
        print("Data was deleted from beginning.")
    
    def delete_from_end(self):
        if self.first is None:
            print("[Error] : List is empty.")
        elif self.first.next is None:
            self.first = self.last = None
        else:
            self.last = self.last.prev
            self.last.next = None
        print("Data was deleted from end.")
    
    def delete_from_specifi_pos(self, pos):
        if pos <= 0:
            print("[Error]: Invalid Position")
        elif pos == 1:
            self.delete_from_beginning()
        else:
            temp = self.first
            for i in range(1, pos-1):
                if temp.next is None:
                    break
                temp = temp.next
            
            if(temp.next == self.last):
                self.delete_from_end()

            tempp = temp.next
            temp.next = tempp.next
            tempp.next.prev = temp
            
        print(f"Data was deleted from {pos}")


dll = DLL()
dll.insert_at_beginning(50)
dll.insert_at_beginning(33)
dll.insert_at_beginning(98)
dll.display()  
dll.insert_at_end(20)
dll.display()  
dll.insert_at_beginning(522)
dll.display()  
dll.insert_at_specific_position(10, 2) 
dll.display()  
dll.delete_from_beginning() 
dll.display()
dll.delete_from_end()        
dll.display()
dll.delete_from_specifi_pos(2)
dll.display()

    
